[FLINK-6281] Create TableSink for JDBC.

haohui opened the pull request:
This PR implements the `StreamTableSink` interface for the JDBC connectors so that the streaming SQL APIs can directly interact with them.

fhueske commented:
Hi @haohui, I think a JdbcTableSink would be a great feature! 

However, there is a big issue with wrapping the `JdbcOutputFormat`. OutputFormats are not integrated with Flink's checkpointing mechanism. The `JdbcOutputFormat` buffers rows to write them out in batches. When records are buffered that arrived before the last checkpoint, they will be lost in case of a failure because they will not be replayed.

The JdbcTableSink should be integrated with Flink's checkpointing mechanism. In a nutshell, it should buffer records and commit them to the database when a checkpoint is taken. I think we need to think a bit more about a proper design for this feature. @zentol and @aljoscha might have some thoughts on this as well as they are more familiar with the implementation of checkpoint-aware sinks.

What do you think?

haohui commented:
Thanks for the review. I'm not aware of the fact that OutputFormats are not integrated with Flink's checkpointing mechanism.

To address this problem, maybe we can do something similar to the `FlinkKafkaProducerBase`? What do you think @zentol and @aljoscha ?

aljoscha commented:
There is `GenericWriteAheadSink`, which buffers elements and writes to Cassandra. I think this needs some more thought, maybe a design outline on the Jira issue. I'm also not 100 % sure if the generic write-ahead sink will work for this. @zentol might have a better answer, though.

zentol commented:
The GenericWriteAheadSink would work for this. It can be implemented just the `CassandraWriteAheadSink`. Since we can also use transactions it can be a bit more sophisticated even.

zentol commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCTableSink.java:
```diff
@@ -0,0 +1,91 @@
+	protected boolean sendValues(Iterable<Row> value, long timestamp) throws Exception {
+		for (Row r : value) {
+			try {
+				outputFormat.writeRecord(r);
```
This doesn't guarantee in any way  that the values are actually being sent; you need some kind of flushing functionality for this to work properly.

zentol commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCTableSink.java:
```diff
@@ -0,0 +1,91 @@
+	private final String[] fieldNames;
+	private final TypeInformation[] fieldTypes;
+
+	public JDBCTableSink(CheckpointCommitter committer, TypeSerializer<Row> serializer,
```
I would propose either adding a JDBCCheckpointCommitter that cooperates with the sink (as seen in this [prototype](https://github.com/zentol/flink/commit/92e878b59a7371ac9cad402d0b009c7439cd1900) or omitting the `CheckpointCommitter` argument and providing a dummy to the `GenericWriteAheadSink`.

haohui commented:
Thanks for your pointer of the prototype!

> Do you intend to provide exactly-once guarantees for arbitrary updates?

As I think about it a little bit more, I think it might make sense to start with the at-least-once semantic first. In practice we make the JDBC call idempotent using `INSERT IF NOT EXISTS`.

The exactly-once part is more tricky and let's separate it out for now. What do you think?

fhueske commented:
I think an at-least-once sink with support for upserts would also be very useful.

zentol commented:
Great, just wanted to make sure we're on the same page.

Why did you revert the usage of the GenericWriteAheadSink? Now we're back to where we started, not having any guarantee that data is written when a checkpoint is being completed.

haohui commented:
Correct me if I'm wrong -- will something like the following work?

```
+	@Override
+	public void snapshotState(FunctionSnapshotContext context) throws Exception {
+		outputFormat.flush();
+	}
```

haohui commented:
@fhueske @zentol can you please take another look? Thanks.

zentol commented on flink-connectors/flink-jdbc/src/test/java/org/apache/flink/api/java/io/jdbc/JDBCTableSinkTest.java:
```diff
@@ -0,0 +1,79 @@
+		sink.emitDataStream(dataStream);
+		sink.snapshotState(mock(FunctionSnapshotContext.class));
+		verify(dataStream).addSink(sink);
+		verify(outputFormat).flush();
```
let's not use mocking for this test. Just create an actual format/sink, give N values to the sink where N < batchSize, verify they haven't been written yet, call flush, verify they were written.

zentol commented on flink-connectors/flink-jdbc/src/test/java/org/apache/flink/api/java/io/jdbc/JDBCTableSinkTest.java:
```diff
@@ -0,0 +1,79 @@
+		@SuppressWarnings("unchecked")
+		DataStream<Row> dataStream = (DataStream<Row>) mock(DataStream.class);
+		sink.emitDataStream(dataStream);
+		verify(dataStream).addSink(sink);
```
you don't have to test this, as it is not a detail of the JDBCTableSink but the table API.

zentol commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCOutputFormat.java:
```diff
@@ -202,14 +202,20 @@ public void writeRecord(Row row) throws IOException {
 
+	void flush() throws SQLException {
+		if (upload != null) {
+			upload.executeBatch();
```
It's been a while since i worked with JDBC, I take it this is a synchronous call? What happens if this call fails?

haohui replied:
It is a synchronous call. It will throw `SQLException` and abort the sink. The behavior has not been changed.

zentol commented:
+1.

fhueske commented:
Thanks @zentol! 
I'll have a look at it as well.

fhueske commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCTableSink.java:
```diff
@@ -0,0 +1,102 @@
+	private String[] fieldNames;
+	private TypeInformation[] fieldTypes;
+
+	public JDBCTableSink(JDBCOutputFormat outputFormat) {
```
What do you think about not exposing the `JDBCOutputFormat` to the user, but to configure it internally.

Of course we would need many of the configuration parameters (user, pw, driver, dburl, and table name). Users could either specify field names of the table to write to (fields are mapped by position) or not (we use the field names of the `Table` to emit). For this information we can construct a parameterized insert query: `INSERT INTO $table ($f1, $f2, $f3) VALUES (?, ?, ?)`.  The field types are automatically provided by the `configure()` call.

This would be a tighter integration with the Table API (using provided field types and possibly field names).
Does this work for your use case or do you need the flexibility of specifying your own query?

fhueske commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCTableSink.java:
```diff
@@ -0,0 +1,102 @@
+/**
+ * An at-least-once Table sink for JDBC.
+ */
+public class JDBCTableSink extends RichSinkFunction<Row>
```
Rename to `JDBCAppendTableSink`?

fhueske replied:
I think it could also be easily extended to support batch output by implementing the `BatchTableSink` interface and implementing `emitDataSet(dataSet: DataSet[Row])` as
```
def emitDataSet(dataSet: DataSet[Row]): Unit = {
  dataSet.output(jdbcOutputFormat)
}
```

haohui commented:
It is important to have some flexibility on the query as different SQL engines have slightly different syntax on DML.

For example, SQLite supports INSERT OVERWRITE where MySQL supports REPLACE INTO to upsert records with unique keys.

I like the APIs you proposed, do you think it addresses your concerns if it forwards the parameters to JDbCOutputFormat and construct it internally?

fhueske commented:
Yes, I think that would work well. We could provide a builder for the table source and forward calls to the `JDBCOutputFormatBuilder`. 

Regarding the query template: we can also make it an optional parameter to override the standard `INSERT INTO ... VALUES` template if set. 

If you want the source to support upsert writes you might want to implement the `UpsertStreamTableSink` rather than the `AppendStreamTableSink`. The `UpsertStreamTableSink` supports append writes as well as updates and deletes by exposing the unique key fields of a `Table`.

fhueske commented on flink-connectors/flink-jdbc/src/test/java/org/apache/flink/api/java/io/jdbc/JDBCTableSinkTest.java:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	@Test
+	public void testFlush() throws Exception {
```
This tests rather the `JdbcOutputFormat.flush()` method. I would move (and adapt) this test to the `JdbcOutputFormatTest`.

For the `JdbcTableSink` we need tests that check the configuration (which is done below) and a check that the `emitDataStream()` method (and the returned `SinkFunction`) is working correctly. I would do this by calling `emitDataStream()` with a mocked `DataStream<Row>` and fetching the `SinkFunction` from the returned `DataStreamSink` (`sink.getTransformation().getOperator().getUserFunction()`).

fhueske commented:
It would also be good to add the `JdbcTableSink` to the documentation (incl. an example) once the API is fixed.

haohui commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCAppendTableSink.java:
```diff
@@ -0,0 +1,85 @@
+
+	@Override
+	public TableSink<Row> configure(String[] fieldNames, TypeInformation<?>[] fieldTypes) {
+		JDBCAppendTableSink copy = new JDBCAppendTableSink(sink.outputFormat);
```
The `JDBCOutputFormat` now is only constructed via `JDBCAppendableSinkBuilder`, thus the types should always match, but it is a good idea to add the checks to catch potential bugs.

fhueske commented on flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCTypeUtil.java:
```diff
@@ -0,0 +1,87 @@
+		m.put(LONG_TYPE_INFO, Types.BIGINT);
+		m.put(FLOAT_TYPE_INFO, Types.FLOAT);
+		m.put(DOUBLE_TYPE_INFO, Types.DOUBLE);
+		m.put(CHAR_TYPE_INFO, Types.SMALLINT);
```
`JDBCOutputFormat` will insert a `SMALLINT` by casting to `short`. This cast will fail for `Character`. 

Please double check the type assignment and align it with `JDBCOutputFormat.writeRecord()`.

fhueske commented on docs/dev/table/sourceSinks.md:
```diff
@@ -202,7 +202,38 @@ val csvTableSource = CsvTableSource
+  .setDrivername("org.apache.derby.jdbc.EmbeddedDriver")
+  .setDBUrl("jdbc:derby:memory:ebookshop")
+  .setQuery("INSERT INTO books (id) VALUES (?)")
+  .setFieldTypes(new TypeInformation<?>[] {INT_TYPE_INFO})
```
change to `setParameterTypes()` if we rename the method.

haohui commented:
Thanks @fhueske ! Updated the PR to address the comments.

fhueske commented:
Thanks for the update @haohui.

I'll merge this PR :-)
I have to change some of the type mappings to make it work with `JDBCOutputFormat`. I think it would be a good idea to redesign the type handling in `JDBCOutputFormat`. I'll open a JIRA for that.

Thanks, Fabian
